File: 11-Binary-Search/python/01_find_nth_root_of_a_number.py
class Solution:
    def NthRoot(self, N: int, M: int) -> int:
        low, high = 1, M

        def compute_power(base):
            res = 1
            for _ in range(N):
                res *= base
                if res > M:
                    return res
            return res

        while low <= high:
            mid = (low + high) // 2
            power = compute_power(mid)

            if power == M:
                return mid
            elif power < M:
                low = mid + 1
            else:
                high = mid - 1

        return -1

File: 11-Binary-Search/cpp/01-find-nth-root-of-a-number.cpp
class Solution {
public:
    int NthRoot(int N, int M) {
        int low = 1, high = M;

        while (low <= high) {
            int mid = low + (high - low) / 2;
            long long power = computePower(mid, N, M);

            if (power == M) {
                return mid;
            } else if (power < M) {
                low = mid + 1;
            } else {
                high = mid - 1;
            }
        }

        return -1;
    }

private:
    long long computePower(int base, int N, int M) {
        long long result = 1;
        for (int i = 0; i < N; i++) {
            result *= base;
            if (result > M) {
                return result;
            }
        }
        return result;
    }
};
